File: __tests__/lib/byok/providers.test.ts
import { SUPPORTED_PROVIDERS, getProvider } from '@/lib/byok/providers'

describe('SUPPORTED_PROVIDERS SSRF allowlist invariants', () => {
  it('identity: every key resolves to a provider whose id matches', () => {
    for (const [id, p] of Object.entries(SUPPORTED_PROVIDERS)) {
      expect(p.id).toBe(id)
      expect(p.name).toBeTruthy()
      expect(p.defaultModel).toBeTruthy()
      expect(p.baseUrl).toBeTruthy()
      expect(getProvider(id)).toBe(p)
    }
  })

  it('proxy providers only ever dial fixed https upstreams — no open relay', () => {
    for (const p of Object.values(SUPPORTED_PROVIDERS)) {
      if (p.kind !== 'proxy') continue
      expect(p.baseUrl.startsWith('https://')).toBe(true)
      // The allowlist must be a literal host, not CLIENT-controlled input:
      // a baseUrl with interpolation/query injection would break the SSRF guard.
      expect(p.baseUrl).not.toMatch(/\$\{/)
      expect(p.baseUrl).not.toMatch(/[{}\s]/)
      // A proxy provider MUST strip an auth header the client could smuggle.
      expect(p.authHeaderName).toBeTruthy()
      expect(p.auth.kind).not.toBe('none')
    }
  })

  it('auth header name matches each provider auth style', () => {
    expect(SUPPORTED_PROVIDERS.anthropic.auth).toEqual({ kind: 'x-api-key' })
    expect(SUPPORTED_PROVIDERS.anthropic.authHeaderName).toBe('x-api-key')
    expect(SUPPORTED_PROVIDERS.gemini.auth).toEqual({ kind: 'x-goog-api-key' })
    expect(SUPPORTED_PROVIDERS.gemini.authHeaderName).toBe('x-goog-api-key')
    for (const id of ['openai', 'nvidia-nim', 'openrouter']) {
      expect(SUPPORTED_PROVIDERS[id]!.auth).toEqual({ kind: 'bearer' })
      expect(SUPPORTED_PROVIDERS[id]!.authHeaderName).toBe('authorization')
    }
  })

  it('anthropic always sends its version header', () => {
    expect(SUPPORTED_PROVIDERS.anthropic.defaultHeaders).toEqual({ 'anthropic-version': '2023-06-01' })
  })

  it('ollama is the only clientDirect provider and routes to localhost', () => {
    const direct = Object.values(SUPPORTED_PROVIDERS).filter(p => p.kind === 'clientDirect')
    expect(direct).toHaveLength(1)
    expect(direct[0]!.id).toBe('ollama')
    expect(SUPPORTED_PROVIDERS.ollama.baseUrl).toMatch(/localhost|127\.0\.0\.1/)
    expect(SUPPORTED_PROVIDERS.ollama.auth).toEqual({ kind: 'none' })
    expect(SUPPORTED_PROVIDERS.ollama.authHeaderName).toBe('')
  })

  it('rejects unknown provider ids', () => {
    expect(getProvider('not-a-provider')).toBeUndefined()
  })
})

File: mobile/fajrak_flutter/test/services/byok/providers_test.dart
import 'package:fajrak/services/byok/providers.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('SUPPORTED_PROVIDERS', () {
    test('has the full web provider list (openai, anthropic, nvidia-nim, openrouter, gemini, ollama)', () {
      expect(
        SUPPORTED_PROVIDERS.keys.toSet(),
        {'openai', 'anthropic', 'nvidia-nim', 'openrouter', 'gemini', 'ollama'},
      );
    });

    test('every entry has a fixed non-empty baseUrl', () {
      for (final entry in SUPPORTED_PROVIDERS.entries) {
        expect(entry.value.baseUrl, isNotEmpty, reason: '${entry.key} baseUrl must be fixed');
      }
    });

    test('exactly one clientDirect provider: ollama', () {
      final clientDirect =
          SUPPORTED_PROVIDERS.values.where((p) => p.kind == 'clientDirect').toList();
      expect(clientDirect, hasLength(1));
      expect(clientDirect.single.id, 'ollama');
    });

    test('proxy providers match the web proxy list', () {
      final proxyIds = SUPPORTED_PROVIDERS.values
          .where((p) => p.kind == 'proxy')
          .map((p) => p.id)
          .toSet();
      expect(proxyIds, {'openai', 'anthropic', 'nvidia-nim', 'openrouter', 'gemini'});
    });

    test('mirror of auth kinds and authHeaderNames', () {
      final cases = {
        'openai': (ByokAuth.bearer, 'authorization'),
        'anthropic': (ByokAuth.xApiKey, 'x-api-key'),
        'nvidia-nim': (ByokAuth.bearer, 'authorization'),
        'openrouter': (ByokAuth.bearer, 'authorization'),
        'gemini': (ByokAuth.xGoogApiKey, 'x-goog-api-key'),
        'ollama': (ByokAuth.none, ''),
      };
      for (final entry in cases.entries) {
        final p = SUPPORTED_PROVIDERS[entry.key];
        expect(p, isNotNull, reason: '${entry.key} must be registered');
        expect(p!.auth, entry.value.$1, reason: '${entry.key} auth kind');
        expect(p.authHeaderName, entry.value.$2, reason: '${entry.key} authHeaderName');
      }
    });

    test('anthropic carries the required version header', () {
      final anthropic = SUPPORTED_PROVIDERS['anthropic'];
      expect(anthropic?.defaultHeaders, {'anthropic-version': '2023-06-01'});
    });

    test('proxy providers send no extra headers except anthropic version header', () {
      for (final p in SUPPORTED_PROVIDERS.values) {
        if (p.kind == 'proxy' && p.id != 'anthropic') {
          expect(p.defaultHeaders, isEmpty, reason: '${p.id} should not send extra headers');
        }
      }
    });
  });

  group('getProvider', () {
    test('returns the registered provider for "openai"', () {
      final p = getProvider('openai');
      expect(p, isNotNull);
      expect(p!.kind, 'proxy');
    });

    test('returns null for an unknown id', () {
      expect(getProvider('unknown'), isNull);
    });
  });
}